libraries/AC_PrecLand/AC_PrecLand_StateMachine.h

```diff
@@ -0,0 +1,107 @@
+    enum class RetryStrictness: uint8_t {
+        NOT_STRICT = 0,         // This is the behaviour on Copter 4.1 and below. The vehicle will land ASAP irrespective of target in sight or not
+        NORMAL,                 // Vehicle will retry a failed prec landing; if the target isn't found, it will land vertically
+        VERY_STRICT             // Same as above, except vehicle will never land if the target isn't found
+    };
+
+    // which retry action should be done
+    enum class RetryAction: uint8_t {
+        GO_TO_LAST_LOC = 0,     // Go to the last location where landing target was detected
+        GO_TO_TARGET_LOC        // Go towards the location of the detected landing target
+    };
+
+private:
+
+    // Target is lost (i.e we had it in sight some time back), this method helps decide on what needs to be done next
+    // The chosen action depends on user set landing strictness and will be returned by this function
+    // Parameters: Vector3f "retry_pos_cm" is filled with the required location if we need to retry landing.
+    Status do_target_lost_actions(Vector3f &retry_pos_cm);
+
+    // Retry landing based on a previously known location of the landing target
+    // Returns the action that should be taken by the vehicle
+    // Vector3f "retry_pos_cm" is filled with the required location.
+    Status retry_landing(Vector3f &retry_pos_cm);
+
+    // Reset the landing statemachines. This needs to be called everytime the landing target is back in sight.
+    // So that if the landing target goes out of sight again, we can start the failed landing procedure back from the beginning stage
+    void reset_failed_landing_statemachines();
+
+    // State machine for action to do when Landing target is lost (after it was in sight a while back)
+    enum class TargetLostAction: uint8_t {
+        INIT = 0,               // Decide on what action needs to be taken
+        DESCEND,                // Descend for sometime (happens if we have just lost the target)
+        LAND_VERTICALLY,        // Land vertically
+        RETRY_LANDING,          // Retry landing (only possible if we had the landing target in sight sometime during the flight)
+    };
+
+    TargetLostAction landing_target_lost_action;  // Current action being done in the Lost Landing target state machine
+
+    // State Machine for landing retry
+    enum class RetryLanding : uint8_t {
+        INIT = 0,               // Init the retry statemachine. This would involve increasing the retry counter (so we how many times we have already retried)
+        IN_PROGRESS,            // Retry in progress, we wait for the vehicle to get close to the target location
+        DESCEND,                // Descend to the original height from where we had started the retry
+        COMPLETE                // Retry completed. We try failsafe measures after this
+    };
+    RetryLanding retry_state;   // Current action being done in the Landing retry state machine
```

Comment:
let's be consistent about whether we use the underscore prefix for the member names.  Either having them or not having them is fine but let's be consistent.